In [1]:
from pandas import Series, DataFrame
import pandas as pd
%pylab inline

Populating the interactive namespace from numpy and matplotlib


__Importing And Reading In The Data File__

The data file is saved as a tsv (tab seperated file). You must set sep='\t' to specify that the data is seperated by tabs. Pandas attempts to guess the datatypes of each column. However, the data file is  large and guessing the datatype is memory demanding. Setting low_memory to false will tell pandas to avoid guessing the datatype so save on memory and processing time

In [2]:
df = pd.read_csv('en.openfoodfacts.org.products.tsv',sep='\t',low_memory=False)
print("This data set contains",len(df),"rows and ",len(df.columns),"columns")

This data set contains 356027 rows and  163 columns


This data set contains 356,027 food products from around the world and 163 characteristics for each product. The data was obtained from Open Food Facts (https://world.openfoodfacts.org/) - a free collaborative database of food products with each product uploaded by users who scan the food products and upload pictures of their nutrition labels. The data set contains a wide variety of information on each product ranging from the ingredients to allergens to nutrition facts, all gathered from the packaging labels. You can view the first 5 rows of the dataset below:

In [3]:
df.head()

code                                                url  \
0  0000000003087  http://world-en.openfoodfacts.org/product/0000...   
1  0000000004530  http://world-en.openfoodfacts.org/product/0000...   
2  0000000004559  http://world-en.openfoodfacts.org/product/0000...   
3  0000000016087  http://world-en.openfoodfacts.org/product/0000...   
4  0000000016094  http://world-en.openfoodfacts.org/product/0000...   

                      creator   created_t      created_datetime  \
0  openfoodfacts-contributors  1474103866  2016-09-17T09:17:46Z   
1             usda-ndb-import  1489069957  2017-03-09T14:32:37Z   
2             usda-ndb-import  1489069957  2017-03-09T14:32:37Z   
3             usda-ndb-import  1489055731  2017-03-09T10:35:31Z   
4             usda-ndb-import  1489055653  2017-03-09T10:34:13Z   

  last_modified_t last_modified_datetime                    product_name  \
0      1474103893   2016-09-17T09:18:13Z              Farine de blé noir   
1      1489069957   2017-03-09T14:32:37Z  Banana Chips Sweetened (Whole)   
2      1489069957   2017-03-09T14:32:37Z                         Peanuts   
3      1489055731   2017-03-09T10:35:31Z          Organic Salted Nut Mix   
4      1489055653   2017-03-09T10:34:13Z                 Organic Polenta   

  generic_name quantity  ... fruits-vegetables-nuts_100g  \
0          NaN      1kg  ...                         NaN   
1          NaN      NaN  ...                         NaN   
2          NaN      NaN  ...                         NaN   
3          NaN      NaN  ...                         NaN   
4          NaN      NaN  ...                         NaN   

  fruits-vegetables-nuts-estimate_100g collagen-meat-protein-ratio_100g  \
0                                  NaN                              NaN   
1                                  NaN                              NaN   
2                                  NaN                              NaN   
3                                  NaN                              NaN   
4                                  NaN                              NaN   

  cocoa_100g chlorophyl_100g carbon-footprint_100g nutrition-score-fr_100g  \
0        NaN             NaN                   NaN                     NaN   
1        NaN             NaN                   NaN                    14.0   
2        NaN             NaN                   NaN                     0.0   
3        NaN             NaN                   NaN                    12.0   
4        NaN             NaN                   NaN                     NaN   

  nutrition-score-uk_100g glycemic-index_100g water-hardness_100g  
0                     NaN                 NaN                 NaN  
1                    14.0                 NaN                 NaN  
2                     0.0                 NaN                 NaN  
3                    12.0                 NaN                 NaN  
4                     NaN                 NaN                 NaN  

[5 rows x 163 columns]

__Selecting Columns of Interest__

This dataset contains a wide variety of characteristics on the products. Below you can observe a snapshot of the columns this data set contains. 

In [4]:
df.columns.values[50:75]

array(['ingredients_that_may_be_from_palm_oil',
       'ingredients_that_may_be_from_palm_oil_tags', 'nutrition_grade_uk',
       'nutrition_grade_fr', 'pnns_groups_1', 'pnns_groups_2', 'states',
       'states_tags', 'states_en', 'main_category', 'main_category_en',
       'image_url', 'image_small_url', 'energy_100g',
       'energy-from-fat_100g', 'fat_100g', 'saturated-fat_100g',
       '-butyric-acid_100g', '-caproic-acid_100g', '-caprylic-acid_100g',
       '-capric-acid_100g', '-lauric-acid_100g', '-myristic-acid_100g',
       '-palmitic-acid_100g', '-stearic-acid_100g'], dtype=object)

As this analysis is focused on gaining a macro level understanding of how different countries eat, let's clean up the dataset to focus on the more generic food product characteristics.

Additionally, any column that ends in '\_en' contains that columns characteristics in English. We will drop any columns that contain similar information in another form. For example, the column 'countries_en' contains the name of the country (in English) where each product was scanned and uploaded. The columns 'countries' and 'countries_tags' contain similar information but in a messier format. We will stick with 'countries_en' for ease of analysis. 

In [3]:
columns_to_drop = ['code', 'url', 'creator', 'created_t', 'created_datetime',
       'last_modified_t', 'last_modified_datetime','no_nutriments', 'ingredients_from_palm_oil_n',
       'ingredients_from_palm_oil', 'ingredients_from_palm_oil_tags',
       'ingredients_that_may_be_from_palm_oil_n',
       'ingredients_that_may_be_from_palm_oil',
       'ingredients_that_may_be_from_palm_oil_tags','pnns_groups_1', 'pnns_groups_2',  '-butyric-acid_100g', '-caproic-acid_100g', '-caprylic-acid_100g',
       '-capric-acid_100g', '-lauric-acid_100g', '-myristic-acid_100g',
       '-palmitic-acid_100g', '-stearic-acid_100g',
       '-arachidic-acid_100g', '-behenic-acid_100g',
       '-lignoceric-acid_100g', '-cerotic-acid_100g',
       '-montanic-acid_100g', '-melissic-acid_100g',
       'monounsaturated-fat_100g', 'polyunsaturated-fat_100g',
       'omega-3-fat_100g', '-alpha-linolenic-acid_100g',
       '-eicosapentaenoic-acid_100g', '-docosahexaenoic-acid_100g',
       'omega-6-fat_100g', '-linoleic-acid_100g',
       '-arachidonic-acid_100g', '-gamma-linolenic-acid_100g',
       '-dihomo-gamma-linolenic-acid_100g', 'omega-9-fat_100g',
       '-oleic-acid_100g', '-elaidic-acid_100g', '-gondoic-acid_100g',
       '-mead-acid_100g', '-erucic-acid_100g', '-nervonic-acid_100g','-maltose_100g', '-maltodextrins_100g',
       'starch_100g', 'polyols_100g',  'casein_100g', 'serum-proteins_100g', 'nucleotides_100g',  'vitamin-b2_100g', 'vitamin-pp_100g', 'vitamin-b6_100g',
       'vitamin-b9_100g', 'folates_100g', 'pantothenic-acid_100g', 'silica_100g',
       'bicarbonate_100g','chloride_100g', 'phosphorus_100g', 'copper_100g', 'manganese_100g', 'fluoride_100g',
       'selenium_100g', 'chromium_100g', 'molybdenum_100g', 'iodine_100g',
       'caffeine_100g', 'taurine_100g', 'ph_100g',
       'fruits-vegetables-nuts_100g',
       'fruits-vegetables-nuts-estimate_100g',
       'collagen-meat-protein-ratio_100g', 'cocoa_100g',
       'chlorophyl_100g','water-hardness_100g', 'categories',
       'categories_tags', 'labels',
       'labels_tags','emb_codes', 'emb_codes_tags','allergens','countries','countries_tags',
        'additives_n', 'additives', 'additives_tags','traces', 'traces_tags', 'states',
       'states_tags', 'main_category']
clean_df = df.drop(columns_to_drop,axis=1)

Our dataset now has the following columns:

In [4]:
clean_df.columns.values

array(['product_name', 'generic_name', 'quantity', 'packaging',
       'packaging_tags', 'brands', 'brands_tags', 'categories_en',
       'origins', 'origins_tags', 'manufacturing_places',
       'manufacturing_places_tags', 'labels_en',
       'first_packaging_code_geo', 'cities', 'cities_tags',
       'purchase_places', 'stores', 'countries_en', 'ingredients_text',
       'allergens_en', 'traces_en', 'serving_size', 'additives_en',
       'nutrition_grade_uk', 'nutrition_grade_fr', 'states_en',
       'main_category_en', 'image_url', 'image_small_url', 'energy_100g',
       'energy-from-fat_100g', 'fat_100g', 'saturated-fat_100g',
       'trans-fat_100g', 'cholesterol_100g', 'carbohydrates_100g',
       'sugars_100g', '-sucrose_100g', '-glucose_100g', '-fructose_100g',
       '-lactose_100g', 'fiber_100g', 'proteins_100g', 'salt_100g',
       'sodium_100g', 'alcohol_100g', 'vitamin-a_100g',
       'beta-carotene_100g', 'vitamin-d_100g', 'vitamin-e_100g',
       'vitamin-k_100g', 'vit

In [5]:
clean_df.head()

product_name generic_name quantity packaging  \
0              Farine de blé noir          NaN      1kg       NaN   
1  Banana Chips Sweetened (Whole)          NaN      NaN       NaN   
2                         Peanuts          NaN      NaN       NaN   
3          Organic Salted Nut Mix          NaN      NaN       NaN   
4                 Organic Polenta          NaN      NaN       NaN   

  packaging_tags           brands      brands_tags categories_en origins  \
0            NaN  Ferme t'y R'nao  ferme-t-y-r-nao           NaN     NaN   
1            NaN              NaN              NaN           NaN     NaN   
2            NaN   Torn & Glasser     torn-glasser           NaN     NaN   
3            NaN        Grizzlies        grizzlies           NaN     NaN   
4            NaN   Bob's Red Mill   bob-s-red-mill           NaN     NaN   

  origins_tags  ... biotin_100g potassium_100g calcium_100g iron_100g  \
0          NaN  ...         NaN            NaN          NaN       NaN   
1          NaN  ...         NaN            NaN        0.000   0.00129   
2          NaN  ...         NaN            NaN        0.071   0.00129   
3          NaN  ...         NaN            NaN        0.143   0.00514   
4          NaN  ...         NaN            NaN          NaN       NaN   

  magnesium_100g zinc_100g carbon-footprint_100g nutrition-score-fr_100g  \
0            NaN       NaN                   NaN                     NaN   
1            NaN       NaN                   NaN                    14.0   
2            NaN       NaN                   NaN                     0.0   
3            NaN       NaN                   NaN                    12.0   
4            NaN       NaN                   NaN                     NaN   

  nutrition-score-uk_100g glycemic-index_100g  
0                     NaN                 NaN  
1                    14.0                 NaN  
2                     0.0                 NaN  
3                    12.0                 NaN  
4                     NaN                 NaN  

[5 rows x 65 columns]

It is obvious that the dataset contains a lot of missing values. As we are interested in understanding how different countries eat, it is essential to know the name of the country in which the product was scanned and uploaded. 

In [7]:
clean_df2 = clean_df.dropna(subset=['countries_en'], how='any')
print("We dropped",len(clean_df)-len(clean_df2),"rows with missing country names")
print("Our new data set has",len(clean_df2),'rows')

We dropped 275 rows with missing country names
Our new data set has 355752 rows


__Where Do Our Food Products Come From?__

The column 'countries_en' contains the name of the country (in English) where each product was scanned and uploaded. We will assume that's where the product was purchased. 

In [8]:
clean_df2['countries_en'].value_counts().sort_values(ascending=False)

United States                      173159
France                             123961
Switzerland                         14932
Germany                              8470
Spain                                5529
                                    ...  
Czech Republic,Hungary,Slovakia         1
France,Saudi Arabia                     1
Belgium,France,World                    1
Brazil,France,Switzerland               1
Hungary,Romania                         1
Name: countries_en, Length: 779, dtype: int64

Looks like some products contain multiple country names. Entries contain multiple countries if they contain a comma in the country name. Let's see how many:

In [9]:
import re
def multiple_country(s):
    return (len(re.findall(',', s))>0)
    #returns true if entry in 'countries_en' contains multiple country names

counts = clean_df2['countries_en'].map(multiple_country) 
len(counts[counts==True]) #number of entries that contain multiple names 

6412

That's quite a few products that have multiple country names listed. As the entry in 'countries_en' contains the country the product was scanned and uploaded to the database in, it means that product is found within that country. Let's duplicate those products so that each product is associated with a single country name.

For example, if a product was found in the Czech Republic, Hungary, and Slovakia, we want to have this product appear 3 times in our data set with a single country name listed in our 'countries_en' column. 

Even though this will result in duplicated products, we are interested in analyzing foods found in each country, not the individual products. It will aid us in comparing how countries eat. 

In [12]:
clean_df2 = clean_df2.assign(country=clean_df2['countries_en'].str.split(',')).explode('country')
#takes the string of countries in 'countries_en'
#splits them on the comma 
#creates a list of countries per product

#the method explode "explodes" the list into separate rows on the country name
#and retains all other information for each row 

In [13]:
countries = clean_df2['country'].value_counts().sort_values(ascending=False)
countries

United States                         173708
France                                129320
Switzerland                            17204
Germany                                 9403
Spain                                   6061
                                       ...  
Other-turquie                              1
Suriname                                   1
fr:Bourgogne-aube-nogent-sur-seine         1
fr:Irland                                  1
Somalia                                    1
Name: country, Length: 224, dtype: int64

Our dataset now contains products from 174 different countries. There are still some weirdly formatted country names in this dataset. As we're interested in comparing how countries eat, we need to look at countries that have enough products. While we may not be sure how many products will give us a significant amount of insight into a country's eating habits, let's just look at countries that have more than 1000 products. 


In [14]:
top_countries = countries[countries>1000]
top_countries

United States     173708
France            129320
Switzerland        17204
Germany             9403
Spain               6061
United Kingdom      5987
Belgium             4088
Australia           2319
Russia              1641
Italy               1632
Canada              1070
Name: country, dtype: int64

In [15]:
final = clean_df2[clean_df2['country'].isin(top_countries.index)] 
#grabbing products from countries who have at least 100 products

In [16]:
print("We have information on",len(final),"products")

We have information on 352433 products


Let's save this dataset so we can use it for all future analysis. 

In [17]:
final.to_csv('Food_Products.csv')